### PR TITLE
New text filters

### PR DESCRIPTION
**Pull request by noahmanger**
## Summary
This implements the new checkbox-generating-text-filters. To do so, it:
- Breaks out `CheckboxFilter` and `TextFilter` into their own components, removing the logic from the base `Filter` class
- `TextFilter` instantiates an instance of `CheckboxFilter` on change (for the first time), and adds checkboxes to it every time thereafter. All of the actual filter logic -- adding tags, modifying the query -- is controlled by the `CheckboxFilter`
- Adds removal logic to the `CheckboxFilter` class so that the generated checkboxes can be destroyed
- Universally changes instances of `this.$body` to `this.$elm` for consistency

## Screenshots
image

## Remaining work
- I know this broke some things with other `input[type="text"]` filters, like the date ranges and financial ranges. I think it makes sense to make these their own class, as they won't be generating checkboxes
- Update tests for new functionality

cc @xtine

### CODE REVIEW

**Comment by noahmanger** [comment]
I still want to add a bunch of tests here, but @xtine feel free to take a look.

**Comment by noahmanger** [comment]
Ok @xtine this is ready for review.

**Comment by noahmanger** [comment]
Requires https://github.com/18F/openFEC-web-app/pull/1540

**Comment by ccostino** [comment]
Wow, really awesome to see all of this, nicely done abstracting out the existing filter pieces and build on them!

File: js/checkbox-filter.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+CheckboxFilter.prototype = Object.create(Filter.prototype);
+CheckboxFilter.constructor = CheckboxFilter;
+
+CheckboxFilter.prototype.handleChange = function(e) {
+  var $input = $(e.target);
+  var id = $input.attr('id');
+  var loadedOnce,
+      eventName;
```
Comment (xtine) [inline js/checkbox-filter.js]: Set each declaration in a new line?

File: js/range-filter.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
+'use strict';
+
+var $ = require('jquery');
+var Filter = require('./filter-base');
+
+function RangeFilter(elm) {
+  Filter.Filter.call(this, elm);
+  this.id = this.$input.attr('id');
+  this.$submit = this.$elm.find('button');
```
Comment (xtine) [inline js/range-filter.js]: Submit button isn't being found because the button element isn't within `.js-filter` in 18F/openfec-web-app#1540

Comment (noahmanger) [inline js/range-filter.js]: Good catch. Changing the way that the submit button is found.

**Comment by xtine** [comment]
Currency range filters working great now. Awesome stuff!

**Comment by ccostino** [comment]
Thank you @noahmanger and @xtine!

**Pull request merged**